# 71 Centuries of Virat Kohli

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [2]:
df = pd.read_csv('../input/virat-kohlis-71-centuries/71 Centuries of Virat Kohli.csv')
df.shape

(71, 15)

In [3]:
df.columns

Index(['Score', 'Out/Not Out', 'Against', 'Batting Order', 'Inn.',
       'Strike Rate', 'Venue', 'Column1', 'H/A', 'Date', 'Result', 'Format',
       'Man of the Match', 'Captain', 'Unnamed: 14'],
      dtype='object')

In [4]:
df.drop('Unnamed: 14',axis=1,inplace=True)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Score             71 non-null     int64  
 1   Out/Not Out       71 non-null     object 
 2   Against           71 non-null     object 
 3   Batting Order     71 non-null     int64  
 4   Inn.              71 non-null     int64  
 5   Strike Rate       44 non-null     float64
 6   Venue             71 non-null     object 
 7   Column1           71 non-null     object 
 8   H/A               71 non-null     object 
 9   Date              71 non-null     object 
 10  Result            71 non-null     object 
 11  Format            71 non-null     object 
 12  Man of the Match  71 non-null     object 
 13  Captain           71 non-null     object 
dtypes: float64(1), int64(3), object(10)
memory usage: 7.9+ KB


In [6]:
df['Date'] =  pd.to_datetime(df['Date'], format='%d-%m-%Y')

In [7]:
df.isnull().sum()

Score                0
Out/Not Out          0
Against              0
Batting Order        0
Inn.                 0
Strike Rate         27
Venue                0
Column1              0
H/A                  0
Date                 0
Result               0
Format               0
Man of the Match     0
Captain              0
dtype: int64

In [8]:
df[df['Strike Rate'].isnull()]

Score Out/Not Out       Against  Batting Order  Inn.  Strike Rate  \
0     116         Out     Australia              6     2          NaN   
1     103         Out   New Zealand              5     2          NaN   
2     103         Out       England              5     2          NaN   
3     107         Out     Australia              5     2          NaN   
4     119         Out  South Africa              4     1          NaN   
5     105     Not Out   New Zealand              4     4          NaN   
6     115         Out     Australia              4     2          NaN   
7     141         Out     Australia              4     4          NaN   
8     169         Out     Australia              4     2          NaN   
9     147         Out     Australia              4     2          NaN   
10    103         Out     Sri Lanka              4     2          NaN   
11    200         Out   West Indies              4     1          NaN   
12    211         Out   New Zealand              4     1          NaN   
13    167         Out       England              4     1          NaN   
14    235         Out       England              4     2          NaN   
15    204         Out    Bangladesh              4     1          NaN   
16    103     Not Out     Sri Lanka              4     3          NaN   
17    104     Not Out     Sri Lanka              4     3          NaN   
18    213         Out     Sri Lanka              4     2          NaN   
19    243         Out     Sri Lanka              4     1          NaN   
20    153         Out  South Africa              4     2          NaN   
21    149         Out       England              4     2          NaN   
22    103         Out       England              4     3          NaN   
23    139         Out   West Indies              4     1          NaN   
24    123         Out     Australia              4     2          NaN   
25    254     Not Out  South Africa              4     1          NaN   
26    136         Out    Bangladesh              4     2          NaN   

                                         Venue         Column1   H/A  \
0                                Adelaide Oval        Adelaide  Away   
1                       M. Chinnaswamy Stadium       Bangalore  Home   
2         Vidarbha Cricket Association Stadium          Nagpur  Home   
3                    M. A. Chidambaram Stadium         Chennai  Home   
4                            Wanderers Stadium    Johannesburg  Away   
5                                Basin Reserve      Wellington  Away   
6                                Adelaide Oval        Adelaide  Away   
7                                Adelaide Oval        Adelaide  Away   
8                     Melbourne Cricket Ground       Melbourne  Away   
9                        Sydney Cricket Ground          Sydney  Away   
10                 Galle International Stadium           Galle  Away   
11                 Sir Vivian Richards Stadium     North Sound  Away   
12                              Holkar Stadium          Indore  Home   
13                    ACA-VDCA Cricket Stadium   Visakhapatnam  Home   
14                            Wankhede Stadium          Mumbai  Home   
15  Rajiv Gandhi International Cricket Stadium       Hyderabad  Home   
16                 Galle International Stadium           Galle  Away   
17                                Eden Gardens         Kolkata  Home   
18        Vidarbha Cricket Association Stadium          Nagpur  Home   
19                     Feroz Shah Kotla Ground           Delhi  Home   
20                             SuperSport Park       Centurion  Away   
21                    Edgbaston Cricket Ground      Birmingham  Away   
22                                Trent Bridge      Nottingham  Away   
23      Saurashtra Cricket Association Stadium          Rajkot  Home   
24                               Perth Stadium           Perth  Away   
25     Maharashtra Cricket Association Stadium            Pune  Home   
26             

**These all are Test Centuries here Strike Rate is not that important.**

In [9]:
df['Strike Rate'] = df['Strike Rate'].fillna(0)

In [10]:
df.isnull().sum()

Score               0
Out/Not Out         0
Against             0
Batting Order       0
Inn.                0
Strike Rate         0
Venue               0
Column1             0
H/A                 0
Date                0
Result              0
Format              0
Man of the Match    0
Captain             0
dtype: int64

In [11]:
df.duplicated().sum()

0

**Statastical Description for ODIs and T20Is**

In [12]:
df[df['Format']!='Test'][['Score','Strike Rate']].describe()

Score  Strike Rate
count   44.000000    44.000000
mean   120.840909   114.019545
std     17.743578    25.257567
min    100.000000    84.900000
25%    107.000000    96.632500
50%    116.500000   108.935000
75%    128.250000   120.787500
max    183.000000   200.000000

*Observations:*

* Avg. Score of Virat Kohli in ODIs/T20Is is 121 in those matches where he have scored Century.
* His maximum Score in white ball cricket is 183 Runs.
* He have Scored Centuries with the avg.strike rate of 144 in limited over format.

**Statastical Distribution for Test Matches**

In [13]:
df[df['Format']=='Test']['Score'].describe()

count     27.000000
mean     150.555556
std       48.817031
min      103.000000
25%      106.000000
50%      139.000000
75%      184.500000
max      254.000000
Name: Score, dtype: float64

*Observations:*
* Virat's Maximum Score in Test cricket is 254
* Avg.score of Virat Kohli in those matches where he reached century is 150

In [14]:
px.pie(values = df.Format.value_counts().values,names=df.Format.value_counts().index,hole=0.5,title="Virat Kohli's Tons")

*Greatest in Every Format: Virat Kohli had Scored Tons in each format 43 in ODIs 27 in TEST and 1 in T20I Cricket*

In [15]:
df['Format_century_no'] = df.groupby(['Format'])['Date'].rank(method='max').astype(int)

**Here we hae Used Rank function because our aim is to find the first and latest century Scored by Virat in each format.**

In [16]:
df[df.Format_century_no ==1].iloc[:,0:-1]

Score Out/Not Out     Against  Batting Order  Inn.  Strike Rate  \
0     116         Out   Australia              6     2         0.00   
27    107         Out   Sri Lanka              4     2        93.85   
70    122     Not Out  Afganistan              1     1       200.00   

                                  Venue    Column1   H/A       Date Result  \
0                         Adelaide Oval   Adelaide  Away 2012-01-24   Lost   
27                         Eden Gardens    Kolkata  Home 2009-12-24    Won   
70  Dubai International Cricket Stadium      Dubai  Away 2022-09-08    Won   

   Format Man of the Match Captain  
0    Test               No      No  
27    ODI               No      No  
70   T20I              Yes      No

**Virat Kohli's First Century in each Format :**

* **Test** : 116, Ind vs Aus (24th Jan 2012) at Adelaide Oval,Australia (India Lost this Match while chasing Kohli was batting at number 6)

* **ODI** : 107(Out, str.rate: 93.85), Ind vs SL (24th Dec 2009) at Eden Garden,Kolkata,India (India won the Match while chasing Kohli was batting at number 4)

* **T20I** : 122(Not Out, str.rate: 200), Ind vs Afg (08th Sep 2022) at Dubai International Stadium,Dubai (India Won this match while defending,Kohli was opener and Virat was awarded with MOM) 


In [17]:
df[(df.Format_century_no ==43)|(df.Format_century_no ==27)|(df.Format_century_no ==1)].iloc[:,:-1].sort_values(by='Date').tail(3)

Score Out/Not Out      Against  Batting Order  Inn.  Strike Rate  \
69    114     Not Out  West Indies              3     2       115.15   
26    136         Out   Bangladesh              4     2         0.00   
70    122     Not Out   Afganistan              1     1       200.00   

                                  Venue         Column1   H/A       Date  \
69                    Queen's Park Oval   Port of Spain  Away 2019-08-14   
26                         Eden Gardens         Kolkata  Home 2019-11-22   
70  Dubai International Cricket Stadium           Dubai  Away 2022-09-08   

   Result Format Man of the Match Captain  
69    Won    ODI              Yes     Yes  
26    Won   Test               No     Yes  
70    Won   T20I              Yes      No

**Virat Kohli's latest Century in each Format :**

* **Test** : 136 (Out), Ind vs Ban (22th Nov 2019) at Eden Garden,Kolkata,India (India won the Match while chasing,Virat batted at number 4 and Kohli was leading India)

* **ODI** : 114 (Not Out, str.rate: 115.15), Ind vs WI (14th Aug 2019) at Queen's Park Oval,Port of Spain (India won the Match while chasing,Virat batted at number 3,Virat was leading the team and was also awarded with Man of the Match)

* **T20I** : 122 (Not Out, str.rate: 200), Ind vs Afg (08th Sep 2022) at Dubai International Stadium,Dubai (India Won this match while defending and Virat was awarded with MOM) 


**Observation :**

*When Virat scored 100 last time India won matches in all 3 Formats*

### Virat's Centuries against other Teams 

In [18]:
Tonvs = df.groupby(['Against','Format','Result'])['Score'].count().reset_index().rename(columns ={'Score':'Number_of_Centuries'})
px.bar(data_frame=Tonvs,x='Against',y='Number_of_Centuries',color ='Format',text_auto=True,pattern_shape='Result')

**Observations :**
* Virat Scored maximum 15 Centuries against the Giant Aussies out of which 8 were in ODIs and 7 were in TESTs which is followed by Sri Lanka(total = 13, ODIs = 8, TESTs= 5).

* Pakistan,Zimbabwe and Afganistan are the teams against whome Virat haven't scored any Test Hundred.

* Virat had Scored his only T20I Century against Afganistan

### Virat's Centuries Over the years

In [19]:
df = df.sort_values(by=['Date'])

In [20]:
l = []
for i in range(1,72):
    l.append(i)

In [21]:
df['Int_Century_count'] = l

In [22]:
fig = px.line(data_frame=df,x='Date',y='Int_Century_count',markers=True,title="Virat's International Hundreds")
fig.update_layout(title_x=0.5)
fig.show()

In [23]:
centuries_by_year = df.groupby([df['Date'].apply(lambda x: x.year),'Format'])['Score'].count().reset_index().rename(columns={'Score':'Number_of_Centuries'})

In [24]:
px.line(data_frame=centuries_by_year,x='Date',y='Number_of_Centuries',color='Format',markers=True,symbol='Format')

In [25]:
df['Date'].value_counts()

2014-12-09    2
2017-08-31    1
2017-12-02    1
2017-11-24    1
2017-11-16    1
             ..
2015-02-15    1
2015-08-12    1
2015-10-22    1
2016-01-17    1
2022-09-08    1
Name: Date, Length: 70, dtype: int64

In [26]:
df[df['Date']=='2014-12-09']

Score Out/Not Out    Against  Batting Order  Inn.  Strike Rate  \
7    141         Out  Australia              4     4          0.0   
6    115         Out  Australia              4     2          0.0   

           Venue    Column1   H/A       Date Result Format Man of the Match  \
7  Adelaide Oval   Adelaide  Away 2014-12-09   Lost   Test               No   
6  Adelaide Oval   Adelaide  Away 2014-12-09   Lost   Test               No   

  Captain  Format_century_no  Int_Century_count  
7     Yes                  8                 28  
6     Yes                  8                 29

**At Adelaide Oval in 2014 against Australia Virat had complited 2 Centuries on Same date i.e. 09th Dec 2014 where he was leading the team but despite of his performance India have to lost the match.**

    - In Second Inning Virat was batting at number 4 and got out on 115
    - In Fourth Inning he was batting at number 4 and got out on 141

### MOM while Scoring Centuries

In [27]:
mom_cen = df.groupby(['Format','Man of the Match'])['Score'].count().reset_index()

In [28]:
fig = px.sunburst(mom_cen,path=['Format','Man of the Match','Score'],values='Score')
fig.update_layout(title='Number of Centuries',title_x=0.5)


**Observtions:**

* **ODIs** : Out of his 43 Centuries in ODI he won MOM 29 times.
* **TESTs** : Out of his 27 Centuries in TEST he won MOM 8 times.
* **T20Is** : He was awarded MOM for his Maiden T20I Ton.

### Win Percent and Virat's Centuries

In [29]:
ton_win = df.groupby(['Result','Format'])['Score'].count().reset_index()

In [30]:
fig = px.sunburst(ton_win,path=['Format','Result'],values='Score')
fig.update_traces(textinfo="label+percent parent")
fig.update_layout(title="Winning and Losing pecentage of India in all three Formats w.r.t Virat's Century",title_x=0.5)
fig.show()

**Observations:**
* India had Won 81% of matches when Virat Scored Century in ODIs
* India had Won 48% and Drawn 26% of matches when Virat Scored Century in TESTs

In [31]:
df_results_against = df.groupby(['Format','Result','Against'])['Score'].count().reset_index()

In [32]:
fig = px.sunburst(df_results_against,path=['Format','Result','Against'],values='Score')
fig.update_traces(textinfo="label+percent parent")

**From the above plot one thing we can clearly observe and that is when ever india lost beside Kohli's Century maximum times opponent was Australia. So though Virat had scored maximum Centuries against Ausssies the Percentage of Winning is not maximum against them**

### Kohli's Centuries Aginst Aussies

In [33]:
fig = px.sunburst(df_results_against[df_results_against['Against']=='Australia'],path=['Format','Result'],values='Score')
fig.update_traces(textinfo="label+percent parent")

**Yes, we were Right beside Kohli's Century India have lost 57% of it's Test matches, 38% of ODIs against Australia.**

### Kohli's Centuries as Captain

In [34]:
ton_cap = df[df['Captain']=='Yes'].groupby(['Captain','Format','Against','Result'])['Score'].count().reset_index()
fig = px.sunburst(ton_cap,path=['Captain','Format','Against','Result'],values='Score')
fig.update_traces(textinfo="label+percent parent")
fig.update_layout(title="Virat's Centuries as a captain and winning Percent of India",title_x=0.5)

**Observation:**

When Virat was leading the team:
* He scored maximum 33% of his Centuries against West Indies in ODIs out of which India won 71% of matches.
* He scored maximum 25% of his Centuries against Sri Lanka in TESTs out of which India had won 40% matches where as 40% were ended in Draw.

In [35]:
df.sample()

Score Out/Not Out    Against  Batting Order  Inn.  Strike Rate  \
42    100     Not Out  Australia              3     2        192.3   

                     Venue  Column1   H/A       Date Result Format  \
42  Sawai Mansingh Stadium   Jaipur  Home 2013-10-16    Won    ODI   

   Man of the Match Captain  Format_century_no  Int_Century_count  
42               No      No                 16                 20

### Kohli's Centuries at Home vs Away

In [36]:
H_a = df.groupby(['Against','H/A'])['Score'].count().reset_index().rename(columns={'Score':'Centuries'})
fig = px.bar(data_frame=H_a,x='Against',y='Centuries',color='H/A',text_auto=True)
fig.update_layout(title = "Virat's Tons at Home and Away from Home", title_x=0.5)
fig.show()

In [37]:
H_a = df.groupby(['Captain','Format','H/A'])['Score'].count().reset_index().rename(columns={'Score':'Centuries'})
fig = px.bar(data_frame=H_a,x='Captain',y='Centuries',color='H/A',pattern_shape='Format',text_auto=True)
fig.update_layout(title = "Virat's Tons at Home and Away from Home as a Captain", title_x=0.5)
fig.show()

# To Be Continued !!!!